proposals/0032-inline-stylesheets.md

```diff
@@ -0,0 +1,89 @@
+
+So far we have not provided a way to allow inlining stylesheets, prompting workarounds. However, coming from other frameworks, users might expect to be able to configure this behavior.
+- SvelteKit allows configuring a threshold under which CSS files will be inlined:  https://kit.svelte.dev/docs/configuration#inlinestylethreshold
+- Fresh lets plugins inject inline styles: https://fresh.deno.dev/docs/concepts/plugins#hook-render
+- Nuxt has a global configuration: https://nuxt.com/docs/api/configuration/nuxt-config#inlinessrstyles
+
+# Goals
+
+- Provide a way to reduce the number of HTTP requests for stylesheets.
+- Maintain current behavior when not explicitly configured.
+- Works with both `server` and `static` outputs.
+
+# Non-Goals
+
+- Identify "critical" CSS rules.
+- Enable a single external stylesheet.
+- Preloading.
+- Inlining of scripts.
+
+# Detailed Design
+
+The decision to inline a stylesheet should be made at build time (`astro build`), based on the `build.inlineStylsheets` and `vite.build.assetsInlineLimit` configuration options. We already use `assetsInlineLimit` elsewhere to decide if a script should be inlined. `build.inlineStylsheets` option could be set to either `"always"`, `"never"`, or `"auto"`. To stay consistent with the current behavior, the default value should be `"never"`. When `build.inlineStylsheets` is set to `"auto"`, `vite.build.assetsInlineLimit` should be respected. If `assetsInlineLimit` is undefined, a default threshold of 4kb should be used.
+
+If the stylesheet is to be inlined, its contents should be added to `PageBuildData` of each page that it is used in. The asset should also be removed from vite's bundle to avoid unnecessary files in the output.
+
+Astro has run into CSS ordering issues and currently ensures that the `<link>` tags are in a specific order. Currently, this order can't be guaranteed when some stylesheets are to be inlined as `<style>` tags: the rendering pipeline receives the two separtely and inserts all the `<link>` tags first. Therefore, an implementation must pass both inline and external stylesheets in the same array or set that preserves the order, and the rendering pipeline should make sure each `SSRElement` gets serialized with the appropriate tag.
+
+# Testing Strategy
+
+An implementation can be tested adequately with fixture-based tests, ensuring the DOM from SSR responses and generated `.html` files has the expected count of style and link tags and that they are placed in the expected order.
+
+# Drawbacks
+
+- Might lead to unexpected results with our current CSS chunking strategy - it tends to create a lot of tiny stylesheets, and each of them might be under the limit, leading to more CSS being inlined than the user might expect.
```

Comment:
I would say that another Drawback of this new option is that it could increase the disparity between the `development` mode and the `production` mode.

Reply:
Dev mode inlines, so this would let you opt into reducing the disparity.

Reply:
If you chose `"auto"`, would that generate a different result from `inlineStylesheets: "never"` and `development`?

Reply:
yes, it would